tests/functional/adapter/utils/test_utils.py

```diff
@@ -0,0 +1,121 @@
+import pytest
+
+from dbt.tests.adapter.utils.test_any_value import BaseAnyValue
+from dbt.tests.adapter.utils.test_bool_or import BaseBoolOr
+from dbt.tests.adapter.utils.test_cast_bool_to_text import BaseCastBoolToText
+from dbt.tests.adapter.utils.test_concat import BaseConcat
+from dbt.tests.adapter.utils.test_dateadd import BaseDateAdd
+from dbt.tests.adapter.utils.test_datediff import BaseDateDiff
+from dbt.tests.adapter.utils.test_date_trunc import BaseDateTrunc
+from dbt.tests.adapter.utils.test_escape_single_quotes import BaseEscapeSingleQuotesQuote
+from dbt.tests.adapter.utils.test_escape_single_quotes import BaseEscapeSingleQuotesBackslash
+from dbt.tests.adapter.utils.test_except import BaseExcept
+from dbt.tests.adapter.utils.test_hash import BaseHash
+from dbt.tests.adapter.utils.test_intersect import BaseIntersect
+from dbt.tests.adapter.utils.test_last_day import BaseLastDay
+from dbt.tests.adapter.utils.test_length import BaseLength
+from dbt.tests.adapter.utils.test_position import BasePosition
+from dbt.tests.adapter.utils.test_replace import BaseReplace
+from dbt.tests.adapter.utils.test_right import BaseRight
+from dbt.tests.adapter.utils.test_safe_cast import BaseSafeCast
+from dbt.tests.adapter.utils.test_split_part import BaseSplitPart
+from dbt.tests.adapter.utils.test_string_literal import BaseStringLiteral
+
+# requires modification
+from dbt.tests.adapter.utils.test_listagg import BaseListagg
+from dbt.tests.adapter.utils.fixture_listagg import models__test_listagg_yml
+from tests.functional.adapter.utils.fixture_listagg import models__test_listagg_no_order_by_sql
+
+
+class TestAnyValue(BaseAnyValue):
+    pass
+
+
+class TestBoolOr(BaseBoolOr):
+    pass
+
+
+class TestCastBoolToText(BaseCastBoolToText):
+    pass
+
+
+@pytest.mark.skip_profile('spark_session')
+class TestConcat(BaseConcat):
+    pass
+
+
+class TestDateAdd(BaseDateAdd):
+    pass
+
+
+@pytest.mark.skip_profile('spark_session')
+class TestDateDiff(BaseDateDiff):
+    pass
```

Comment:
Why are the tests for `datediff` skipped for `spark_session`?

Time logic is complicated, and we have a highly custom implementation, so it feels crucial to test it if at all possible.

On the other hand, I think this logic has been battle-tested for a couple years, which gives a nice vote of confidence.

I'd also like to re-review the BaseDateDiff implementation to see if it has coverage of each possible `datepart`.

Would be awesome if we could find a robust suite of a test cases that cover well-known edge cases like timestamps with a non-00 UTC offset, daylight savings boundaries, leap years, leap seconds, etc.

Spark changed its Julian vs. (Proleptic) Gregorian calendar handling between Spark 2.4 and 3.0, but not sure if we need to worry about that piece at all ([talk](https://databricks.com/session_eu20/comprehensive-view-on-date-time-apis-of-apache-spark-3-0), [slides](https://www.slideshare.net/databricks/comprehensive-view-on-datetime-apis-of-apache-spark-30)).

Reply:
> Would be awesome if we could find a robust suite of a test cases that cover well-known edge cases like timestamps with a non-00 UTC offset, daylight savings boundaries, leap years, leap seconds, etc.

Agreed. This feels important for our work around foundational data types + `current_timestamp` as well.

A bunch of tests have been failing for `spark_session`. I'm skipping them now for expediency, but these tests _are_ running on the four other connection types for which we have testing. Those connection types are:
- `thrift` for local (self-hosted) Spark
- Databricks interactive cluster via HTTP
- Databricks interactive cluster via ODBC
- Databricks SQL endpoint via ODBC

I'll be the first to admit that the `spark_session` connection method is an advanced capability that I don't know lots about :) It's useful for advanced users / PySpark superusers when testing locally, but I would not recommend folks to use in production. It will never be supported in dbt Cloud. We've documented it [as such](https://docs.getdbt.com/reference/warehouse-profiles/spark-profile).

It's true that the `datediff` (and `datedd`) macros produce a LUDICROUS amount of compiled SQL. The alternatives to tons of repeated code would be to run each snippet as an introspective query, store its result, and template it into the subsequent operation.

That talk about calendar switching is ... cool!

I'm sure there are edge cases with these implementations. The work involved in reaching parity / consistency for just the integration test cases that we already have in place was immense. I think that's the feature, for now..?